libraries/AP_BattMonitor/AP_BattMonitor_Scripting.cpp

```diff
@@ -0,0 +1,152 @@
+#include "AP_BattMonitor_Scripting.h"
+
+#if HAL_BATTMON_SCRIPTING_ENABLED
+
+AP_BattMonitor_Scripting::AP_BattMonitor_Scripting(AP_BattMonitor &mon,
+                                                   AP_BattMonitor::BattMonitor_State &mon_state,
+                                                   AP_BattMonitor_Params &params) :
+    AP_BattMonitor_Backend(mon, mon_state, params)
+{
+}
+
+bool AP_BattMonitor_Scripting::capacity_remaining_pct(uint8_t &percentage) const
+{
+    if (!_has_capacity_remaining_pct) {
+        // Fall back to default implementation
+        return AP_BattMonitor_Backend::capacity_remaining_pct(percentage);
+    }
+    percentage = internal_state.capacity_remaining_pct;
+    return true;
+}
+
+bool AP_BattMonitor_Scripting::get_cycle_count(uint16_t &cycles) const
+{
+    if (!_has_cycle_count) {
+        return false;
+    }
+    cycles = internal_state.cycle_count;
+    return true;
+}
+
+// Called by frontend to update the state. Called at 10Hz
+void AP_BattMonitor_Scripting::read() {
+    
+    // Check for timeout, to prevent a faulty script from appearing healthy
+    if (AP_HAL::micros() - internal_state.last_update_us > AP_BATTMONITOR_SCRIPTING_TIMEOUT_MICROS) {
+        _state.healthy = false;
+        return;
+    }
+
+    // Check if the internal state has changed
+    if (internal_state.last_update_us == _state.last_time_micros) {
+        return;
+    }
+
+    WITH_SEMAPHORE(_sem);
+
+    // Update the flags
+    _has_time_remaining = internal_state.time_remaining != UINT32_MAX;
+    _has_consumed_energy = !isnan(internal_state.consumed_wh);
+    _has_current = !isnan(internal_state.current_amps);
+    _has_cell_voltages = internal_state.cell_count > 0;
+    _has_temperature = !isnan(internal_state.temperature);
+    _has_capacity_remaining_pct = internal_state.capacity_remaining_pct != UINT8_MAX;
+    _has_cycle_count = internal_state.cycle_count != UINT16_MAX;
+
+    // If the script has set the cell count, fill in the cell voltages up to that
+    for(int i = 0; i < internal_state.cell_count; i++) {
```

Comment:
```suggestion
    for (uint8_t i = 0; i < MIN(ARRAY_SIZE(cell_pkt.cell_voltages),internal_state.cell_count); i++) {
```
The script should never be able to set a too big cell count, but better to be sure rather read past the end.